File: project/app/src/main/java/org/owntracks/android/model/messages/MessageCmd.java
package org.owntracks.android.model.messages;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

import org.owntracks.android.model.CommandAction;
import org.owntracks.android.support.Preferences;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.EXTERNAL_PROPERTY, property = "_type")
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_EMPTY)
public class MessageCmd extends MessageBase {
    static final String TYPE = "cmd";
    private static final String BASETOPIC_SUFFIX = "/cmd";
    private CommandAction action;

    private MessageWaypoints waypoints;
    private MessageConfiguration configuration;

    @Override
    @NonNull
    public String getBaseTopicSuffix() {
        return BASETOPIC_SUFFIX;
    }

    @Nullable
    public CommandAction getAction() {
        return action;
    }

    public void setAction(CommandAction action) {
        this.action = action;
    }

    @Nullable
    public MessageWaypoints getWaypoints() {
        return waypoints;
    }

    public void setWaypoints(MessageWaypoints m) {
        this.waypoints = m;
    }

    @Nullable
    public MessageConfiguration getConfiguration() {
        return configuration;
    }

    public void setConfiguration(MessageConfiguration m) {
        this.configuration = m;
    }

    @Override
    public boolean isValidMessage() {
        return super.isValidMessage() && (action != null);
    }

    @Override
    public void addMqttPreferences(Preferences preferences) {
        setTopic(preferences.getPubTopicCommands());
    }

    @Override
    @JsonIgnore
    public void setTopic(String topic) {
        // Full topic is needed instead of the normalized base topic to verify if the message arrived on the correct topic
        this._topic = topic;
    }
}

File: project/app/src/main/java/org/owntracks/android/model/messages/MessageCmd.kt
package org.owntracks.android.model.messages

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import org.owntracks.android.model.CommandAction
import org.owntracks.android.preferences.Preferences

@Serializable
@SerialName(MessageCmd.TYPE)
class MessageCmd(@Transient private val messageWithId: MessageWithId = MessageWithRandomId()) :
    MessageBase(), MessageWithId {
  @kotlinx.serialization.EncodeDefault(kotlinx.serialization.EncodeDefault.Mode.ALWAYS)
  @SerialName("_id")
  override var messageId: MessageId = messageWithId.messageId
  var action: CommandAction? = null
  var waypoints: MessageWaypoints? = null
  var configuration: MessageConfiguration? = null

  override fun isValidMessage(): Boolean {
    return super.isValidMessage() && action != null
  }

  override fun toString(): String = "[MessageCmd action=$action]"

  override fun annotateFromPreferences(preferences: Preferences) {}

  override val baseTopicSuffix: String
    get() = BASETOPIC_SUFFIX

  companion object {
    const val TYPE = "cmd"
    private const val BASETOPIC_SUFFIX = "/cmd"
  }
}
